Unset with nil, fix performance by removing default limit=0, convenience methods removed? persisted? and _id!

atlantis opened the pull request:
Sorry this is a bunch of fairly unrelated changes... there are a few things I'm not sure about, so please let me know any modifications you'd like made and I'll be happy happy to make them!

1. in my use-case, I often need to unset fields in mongo (e.g. user timezone is a string or nil to inherit from account timezone... if the user chooses Default I need to wipe the field from the record), so this adds support to do that by `user.timezone = nil` will issue an $unset command
2. because 0 != nil, a limit=0 was being passed to find operations, which apparently caused the batch_size to be set to 0 as well (no records returned initially and then one at time)... i discovered this cause it took several seconds to pull 240 records, and it turned out it was issuing ~240 Mongo::Commands::GetMore commands
3. BSON::ObjectId support in find_by_id (and prevent BSON::ObjectId.new("") because that led to weird data corruption in queries)
4. ability to override the spec db url with ENV["MONGO_URL"]
5. inserted? removed? persisted? and _id! methods on MongoBase
6. respect BSON::Field ignore:true annotations (just like the BSON library does)

atlantis committed:
Add ability to run specs with connection string in MONGO_URL environment variable

atlantis committed:
Respect BSON::Field ignore annotation

Also add convenience methods _id! and id!

atlantis committed:
Update cryomongo version

atlantis committed:
Meh nillable ObjectIDs are convenient but not very Crystal

atlantis committed:
Whoops that function was me not understanding how cursors work

atlantis committed:
nil properties now unset mongo fields

Also, don't want to break the unit tests so back to nillable finder ids.

Finally, quashed a tricky bug where batch_size: 0 was always getting passed through and thus causing n+1 Mongo::Commands::GetMore (took 5 seconds to load ~250 rows lol).

atlantis committed:
Add ability to set fields to nil to unset them in the db

Also avoid hardcoding spec override url - just fall through to the default.

atlantis committed:
Rename peristed_ever? to inserted?

atlantis commented:
Upon further reflection, if you wanted to make to_bson capable of generating the nils needed by unsets_to_bson, you'd have to add like a `reverse_emit_nil` option rather than a `force_emit_nil` option (since we need it exactly flipped, not always emitted).

atlantis committed:
Whoops that's supposed to be protected i suppose

atlantis committed:
Whoops blank id strings lead to errors (and corrupted data)?

Model.find_by_id("") resulted in:

Exception: Invalid utf-8 encoding: invalid bson type in element with field name '��s3�x+?�sclient_id' in object with unknown _id (Exception)
app_1            |   from Exception::CallStack::unwind:Array(Pointer(Void))
app_1            |   from Exception::CallStack#initialize:Array(Pointer(Void))
app_1            |   from Exception::CallStack::new:Exception::CallStack
app_1            |   from raise<Exception>:NoReturn
app_1            |   from ???
app_1            |   from BSON::Decoder
app_1            |   from BSON::Decoder
app_1            |   from BSON::Decoder
app_1            |   from BSON#fetch<String>:Tuple(BSON | BSON::Code | BSON::DBPointer | BSON::Decimal128 | BSON::MaxKey | BSON::MinKey | BSON::ObjectId | BSON::Symbol | BSON::Timestamp | BSON::Undefined | Bool | Float64 | Int32 | Int64 | Regex | Slice(UInt8) | String | Time | UUID | Nil, Bool)
app_1            |   from BSON#[]?<String>:(BSON | BSON::Code | BSON::DBPointer | BSON::Decimal128 | BSON::MaxKey | BSON::MinKey | BSON::ObjectId | BSON::Symbol | BSON::Timestamp | BSON::Undefined | Bool | Float64 | Int32 | Int64 | Regex | Slice(UInt8) | String | Time | UUID | Nil)
app_1            |   from Mongo::Messages::OpMsg#validate:(Mongo::Error+ | Nil)
...

I guess BSON::ObjectID.new("") generates invalid data somehow?

atlantis committed:
Refactoring

Whoops my bad - id! already existed in MongoBase and lots of stuff i added was in the wrong class (Document vs MongoBase vs Collection).  Attempting to fix.

elbywan commented on src/models/database/methods/get.cr:
```diff
@@ -66,7 +66,7 @@ module Moongoon::Traits::Database::Methods::Get
     # NOTE: Other arguments are available but will not be documented here.
     # For more details check out the underlying [`cryomongo`](https://github.com/elbywan/cryomongo) driver documentation and code.
-    def self.find(query = BSON.new, order_by = { _id: -1 }, fields = @@default_fields, skip = 0, limit = 0, **args) : Array(self)
+    def self.find(query = BSON.new, order_by : NamedTuple = { _id: -1 }, fields = @@default_fields, skip = 0, limit : Int32? = nil, **args) : Array(self)
```
I see some issues restricting the type to `NamedTuple`, `order_by` could previously be a dynamically generated `Hash` for instance.

atlantis replied:
Odd... I was having trouble earlier getting it to compile without a named type here, but seems fine now so yay for flexibility 🤷

elbywan commented on src/models/database/internal.cr:
```diff
@@ -16,27 +16,28 @@ module Moongoon::Traits::Database::Internal
       pipeline << BSON.new({"$project": BSON.new(fields)})
     end
     if order_by
+      {{ debug() }}
```
Should be removed 😉.

atlantis replied:
Whoops definitely!

elbywan commented on src/models/models.cr:
```diff
@@ -56,7 +61,10 @@ module Moongoon
+        {% ann = ivar.annotation(::BSON::Field) %}
+        {% unless ann && ann[:ignore] %}
+          "{{ ivar.name }}": self.{{ ivar.name }},
+        {% end %}
```
I'm not sure about this one. 

The `ignore: true` property is already taken into account in the `BSON`specific constructor (`def initialize(bson : BSON)`), and changing the generic constructor would be a breaking change - and it would go against my own usage of the library.

atlantis replied:
Hmm I can definitely see how it would be useful to set arbitrary (non-mongo) properties on objects... pre-caching calculated values in bulk pulls and avoiding n+1 use cases comes to mind.  

My problem is that I have lots of private/protected instance variables without setters which are getting iterated over by `@type.instance_vars`, so calling `instance.{{ivar.id}} = {{ default_value }}` on a private variable with no setter gets me a compile error, and I'd like avoid exposing private variables unnecessarily.

I agree that calling the setter is much better than setting the instance variable directly for a number of reasons (e.g. security and preserving encapsulation).  

So can you think of a way to keep the current flexible behavior but allow folks to use private/protected instance variables on their models?  Seems like a pretty common use case :)

atlantis replied:
Oh yes a good example of the problem would be that `@removed = true` variable... i'd need to make a `private def removed=(val)` function to make it compile (and for every other private instance variable seems like).  But even if I did that, that would still be allowing the hash that's passed into the initializer to set private variables...

atlantis replied:
Hey so I'm happy to nuke this change for now.  I have some crazy ideas on how to check for public setter methods in the constructor and decide what to set that way - but that's a conversation for another pull request!  So for now i'll remove this change from both self.new and to_tuple.  Any other changes need to be made on this pull request?  Otherwise I'll move on to the next one!

elbywan replied:
First, sorry about the time I took to reply I am a bit busy these days!

> calling instance.{{ivar.id}} = {{ default_value }} on a private variable with no setter gets me a compile error

> So can you think of a way to keep the current flexible behavior but allow folks to use private/protected instance variables on their models? Seems like a pretty common use case :)

Ah I understand the issue now and it is definitely an issue to not being able to compile in this case.

>  I have some crazy ideas on how to check for public setter methods in the constructor and decide what to set that way - but that's a conversation for another pull request! 

Allright let's discuss in a separate PR then, I might have some ideas on my own about this 😉.

elbywan replied:
I opened PR #3 that could potentially solve the problem.

elbywan commented on src/models/database/internal.cr:
```diff
@@ -4,7 +4,7 @@ module Moongoon::Traits::Database::Internal
   # Query builders #
 
-  protected def self.format_aggregation(query, stages, fields = nil, order_by = nil, skip = 0, limit = 0)
+  protected def self.format_aggregation(query, stages, fields = nil, order_by = nil, skip = 0, limit : Int32? = 0)
```
`limit : Int32?` would it be possible to change that to a more generic type: `Int?`

atlantis replied:
Done

elbywan commented on src/models/database/methods/patch.cr:
```diff
@@ -66,10 +66,16 @@ module Moongoon::Traits::Database::Methods::Patch
+      changes = BSON.new({
+        "$set": self.to_bson
+      })
+      if unsets = self.unsets_to_bson
```
I don't think that this should be the default behaviour.

For instance I am using partial application with `nil` fields to patch a collection with partial data, and I would rather like `nil` values to not be emitted **and** be ignored instead of being `$unset` (which is the current default behaviour).

Maybe you can put an argument (something like `unset_nil_values : Bool = false`)  to put the feature behind a flag?

atlantis replied:
Ok but remembering to do that every update would be a bit of drag, so i'm adding a global config container for your consideration :) And I'd argue that the standard ORM practice would be to update the data store to reflect nulls set in the object layer (e.g. ruby's ActiveRecord), but as long as there's a way to get either behavior globally it's nbd either way what the default is... thanks again for all your work on this library!

elbywan replied:
> And I'd argue that the standard ORM practice would be to update the data store to reflect nulls set in the object layer (e.g. ruby's ActiveRecord)

I agree, and it's done in the current version by setting `emit_nil_values` to true at the BSON level. I'm just against the `emit_null_values` to false + `$unset` combo by default. 

Thanks for putting it behind a config flag 👍 .

elbywan commented:
Hi @atlantis, thanks for contributing 👍 !

Just finished reviewing the PR, made some comments in there.

atlantis committed:
Remove debugging and keep as much type flexibiltiy as possible

atlantis committed:
Hide unset nil setting behind global config option

atlantis committed:
One more Int32 cleanup

Not super happy with the naming/style of my change but tried not_nil! and that was pretty ugly too (and not_nil! seems to be pretty rare in the codebase).

atlantis committed:
For consistency, make limit default to nil in format_aggregation too

atlantis committed:
Change config to configure... more verby?

atlantis committed:
Don't respect BSON::Field :ignore property in constructor and to_tuple

elbywan commented:
All good now, thanks! 👍
